hand-eye-calibration/full_hand_eye_sample.py

```diff
@@ -0,0 +1,486 @@
+
+    Args:
+        datastring: String of text from .yaml file
+
+    Returns:
+        pose_matrix: Robotic pose as q_r zivid Pose class
+    """
+
+    string = datastring.split("data:")[-1].strip().strip("[").strip("]")
+    pose_matrix = np.fromstring(string, dtype=np.float, count=16, sep=",").reshape(
+        (4, 4)
+    )
+    return zivid.handeye.Pose(pose_matrix)
+
+
+def _save_hand_eye_results(save_dir: Path, transform: np.array, residuals: list):
+    """Save transformation and residuals to folder
+
+    Args:
+        save_dir: Path to where data will be saved
+        transform: 4x4 transformation matrix
+        residuals: List of residuals
+    """
+
+    file_storage_transform = cv2.FileStorage(
+        str(save_dir / f"transformation.yaml"), cv2.FILE_STORAGE_WRITE
+    )
+    file_storage_transform.write("PoseState", transform)
+    file_storage_transform.release()
+
+    file_storage_residuals = cv2.FileStorage(
+        str(save_dir / f"residuals.yaml"), cv2.FILE_STORAGE_WRITE
+    )
+    residual_list = []
+    for res in residuals:
+        tmp = list([res.translation, res.translation])
+        residual_list.append(tmp)
+
+    file_storage_residuals.write(
+        "Per pose residuals for rotation in deg and translation in mm",
+        np.array(residual_list),
+    )
+    file_storage_residuals.release()
+
+
+def _generate_dataset(con: rtde, input_data):
+    """ Generate dataset based on predefined robot poses
+
+    Args:
+        con: Connection between computer and robot
+        input_data: Input package containing the specific input data registers
+
+    returns:
+        save_dir: Location where dataset is saved
+
+    Universal Robot registers:
+        output_int_register_24 = image_count:
+            Counter from robot side of number of images and poses taken.
+        output_bit_register_64 = start_capture:
+            Bool that trigger camera to capture an image
+
+        Signals to UR robot:
+        input_bit_register_64 = finish_capture:
+            Bool that trigger UR robot to move to next position.
+        input_bit_register_65 = ready_to_capture:
+            Bool that tells UR robot that the camera is ready to be used.
+    """
+
+    with zivid.Application() as app:
+        with app.connect_camera() as cam:
+
+            _set_camera_settings(cam)
+            ready_to_capture = True
+
+            save_dir = _generate_folder()
+
+            # Signal robot that camera is ready to capture
+            _write_robot_state(
+                con, input_data, finish_capture=False, camera_ready=ready_to_capture
+            )
+            robot_state = con.receive()
+
+            print(
+                "Initial output robot_states: \n"
+                f"Image count: {robot_state.output_int_register_24} \n"
+                f"Start capture: {robot_state.output_bit_register_64}"
+            )
+
+            images_captured = 1
+            while robot_state.output_int_register_24 != -1:
+
```

Comment:
Consider putting the contents of this loop in a separate function, for readability.

Reply:
done